diff --git a/assembler.py b/assembler.py
index 3d329c4..ed38aed 100644
--- a/assembler.py
+++ b/assembler.py
@@ -6,24 +6,13 @@ NUMBER_OF_NOPS_AFTER_MATRIX_OPERATION = 16
 
 
 operation_name_to_opcode = {
-    "nop": "00",
-    "tensor_core": "01",
-    "burst": "10",
-    "reset": "11"
-}
-
-
-tensor_core_opselects = {
-    "matrix_multiply": "00",
-}
-
-# memory_controller_opselects = {
-#     "matrix_add": "10"
-# }
-
-
-burst_store_load_selects = {
-    "store_and_load": "10"
+    "nop": "000",
+    "tensor_core": "001",
+    "burst": "010",
+    "reset": "011",
+    "matrix_add": "100",
+    "matrix_scale": "101",
+    "matrix_relu": "110"
 }
 
 
@@ -82,43 +71,26 @@ def main():
         store_or_load = None
         
         current_machine_code_line = ""
-        current_machine_code_line += "0" * 48
 
 
         # PARSE THE DIFFERENT INSTRUCTION FORMATS
         if (operation_name in ["nop", "reset"]):
-            current_machine_code_line += "0"*14
+            current_machine_code_line += "0" * 48
+            
+            current_machine_code_line += "0"*13
             current_machine_code_line += operation_name_to_opcode[operation_name]
     
-            
-        elif (operation_name in ["matrix_multiply"]):
-            
-            current_machine_code_line += "0"*12
-            current_machine_code_line += tensor_core_opselects[operation_name]
-            current_machine_code_line += operation_name_to_opcode["tensor_core"]
-            
-            should_have_matrix_operation_nops_after = True
-        
         
         elif (operation_name == "burst"):
+            current_machine_code_line += "0" * 48
             
             store_or_load = assembly_code_tokens[1]
 
             if store_or_load != "store_and_load":
                 raise Exception(f"Only accepts store_and_load as an acceptable argument on line {line_index}")
 
-            current_machine_code_line += "0"*4
+            current_machine_code_line += "0"*13
             
-            # choose the closest power of 2 to the scale factor 
-            # so that we can efficiently scale the matrix being stored in memory
-            if (len(assembly_code_tokens) == 6):
-                scale_factor = math.log2(int(assembly_code_tokens[5]))
-            else:
-                scale_factor = 0
-                
-            current_machine_code_line += number_into_signed_kbit_binary(scale_factor, k=8)
-            
-            current_machine_code_line += burst_store_load_selects[store_or_load]
             current_machine_code_line += operation_name_to_opcode["burst"]
             
             burst_store_address = int(assembly_code_tokens[2])
@@ -127,6 +99,54 @@ def main():
             
             if burst_store_address in [burst_load_matrix1_address, burst_load_matrix2_address]:
                 raise Exception(f"Pipeline Hazard on line {line_index}: Cannot store to and load from the same matrix address ({burst_store_address}) in the same instruction.")
+        
+        
+                    
+        elif (operation_name in ["matrix_multiply"]):
+            current_machine_code_line += "0" * 48
+            
+            current_machine_code_line += "0"*13
+            current_machine_code_line += operation_name_to_opcode["tensor_core"]
+            
+            should_have_matrix_operation_nops_after = True
+
+        
+        elif (operation_name == "matrix_add"):
+            matrix_input1_address = int(assembly_code_tokens[2])
+            matrix_input2_address = int(assembly_code_tokens[3])
+            matrix_output_address = int(assembly_code_tokens[4])
+            
+            current_machine_code_line += number_into_unsigned_kbit_binary(matrix_output_address, k=16)
+            current_machine_code_line += number_into_unsigned_kbit_binary(matrix_input2_address, k=16)
+            current_machine_code_line += number_into_unsigned_kbit_binary(matrix_input1_address, k=16)
+
+            current_machine_code_line += "0"*13
+            current_machine_code_line += operation_name_to_opcode["matrix_add"]
+                    
+        
+        elif (operation_name == "matrix_scale"):
+            current_machine_code_line += "0" * 32
+            
+            matrix_address = int(assembly_code_tokens[2])
+            
+            scale_factor = int(round(math.log2(assembly_code_tokens[3])))
+            
+            current_machine_code_line += number_into_unsigned_kbit_binary(matrix_address, k=16)
+            
+            current_machine_code_line += "0"*5
+            current_machine_code_line += number_into_signed_kbit_binary(scale_factor, k=8)
+            current_machine_code_line += operation_name_to_opcode["matrix_scale"]
+            
+            
+        elif (operation_name == "matrix_relu"):
+            current_machine_code_line += "0" * 32
+            
+            matrix_address = int(assembly_code_tokens[2])
+            
+            current_machine_code_line += number_into_unsigned_kbit_binary(matrix_address, k=16)
+            current_machine_code_line += "0"*13
+            current_machine_code_line += operation_name_to_opcode["matrix_relu"]
+        
             
         else:
             raise Exception(f"operation name {operation_name} on line {line_index} is not supported")
@@ -160,32 +180,22 @@ def main():
                 else:
                     current_machine_code_line += "0"*16
 
-
                 if 2*index+1 < 9:
                     current_machine_code_line += number_into_unsigned_kbit_binary(9*burst_load_matrix1_address + 2*index+1, k=16)
                 else:
                     current_machine_code_line += number_into_unsigned_kbit_binary(9*burst_load_matrix2_address + 2*index+1 - 9, k=16)
                 
-                
                 if 2*index < 9:
                     current_machine_code_line += number_into_unsigned_kbit_binary(9*burst_load_matrix1_address + 2*index,   k=16)
                 else:
                     current_machine_code_line += number_into_unsigned_kbit_binary(9*burst_load_matrix2_address + 2*index - 9, k=16)
                 
+
+                current_machine_code_line += "1"*2
+                current_machine_code_line += "0"*11
             
-                current_machine_code_line += "1"
-                
-                if burst_store_address is not None:
-                    current_machine_code_line += "1"
-                else:
-                    current_machine_code_line += "0"
-                    
-                current_machine_code_line += "0"*10
-            
-                current_machine_code_line += burst_store_load_selects[store_or_load]
                 current_machine_code_line += operation_name_to_opcode["burst"]
-                
-                
+
                 machine_code.append(format(int(current_machine_code_line, 2), "016X") + '\n')
 
 
@@ -200,20 +210,14 @@ def main():
                     
                     current_machine_code_line += "0"
                     current_machine_code_line += "1"
-                    current_machine_code_line += "0"*10
+                    current_machine_code_line += "0"*11
                 
-                    current_machine_code_line += burst_store_load_selects[store_or_load]
                     current_machine_code_line += operation_name_to_opcode["burst"]
-                    
                                     
                     machine_code.append(format(int(current_machine_code_line, 2), "016X") + '\n')
 
+
                 machine_code.append(format(int("0"*64, 2), "016X") + '\n')
-                
-            # else:
-            #     machine_code.append(format(int("0"*64, 2), "016X") + '\n')
-            #     machine_code.append(format(int("0"*64, 2), "016X") + '\n')
-            
             
     
 
diff --git a/assembly_code.asm b/assembly_code.asm
index dfd1301..1936352 100644
--- a/assembly_code.asm
+++ b/assembly_code.asm
@@ -1,171 +1,5 @@
 reset
 nop
-burst store_and_load 19 15 2 
+burst store_and_load 2 0 1
 matrix_multiply
-burst store_and_load 4 19 19
-burst store_and_load 19 15 17
-nop
-matrix_multiply
-burst store_and_load 6 19 19
-burst store_and_load 19 17 6
-matrix_multiply
-burst store_and_load 11 19 19
-burst store_and_load 19 14 16
-matrix_multiply
-burst store_and_load 7 19 19
-burst store_and_load 19 2 10
-nop
-matrix_multiply
-burst store_and_load 18 19 19
-burst store_and_load 19 7 0
-nop
-matrix_multiply
-burst store_and_load 7 19 19
-burst store_and_load 19 2 1
-matrix_multiply
-burst store_and_load 16 19 19
-burst store_and_load 19 7 8
-matrix_multiply
-burst store_and_load 17 19 19
-burst store_and_load 19 4 18
-matrix_multiply
-burst store_and_load 15 19 19
-burst store_and_load 19 13 6
-nop
-matrix_multiply
-burst store_and_load 11 19 19
-burst store_and_load 19 13 13
-matrix_multiply
-burst store_and_load 3 19 19
-burst store_and_load 19 1 12
-matrix_multiply
-burst store_and_load 7 19 19
-burst store_and_load 19 6 6
-matrix_multiply
-burst store_and_load 13 19 19
-burst store_and_load 19 5 8
-matrix_multiply
-burst store_and_load 14 19 19
-burst store_and_load 19 17 3
-nop
-matrix_multiply
-burst store_and_load 2 19 19
-burst store_and_load 19 7 5
-matrix_multiply
-burst store_and_load 6 19 19
-burst store_and_load 19 12 1
-nop
-matrix_multiply
-burst store_and_load 12 19 19
-burst store_and_load 19 8 14
-nop
-matrix_multiply
-burst store_and_load 4 19 19
-burst store_and_load 19 6 9
-nop
-matrix_multiply
-burst store_and_load 18 19 19
-burst store_and_load 19 17 1
-matrix_multiply
-burst store_and_load 1 19 19
-burst store_and_load 19 18 15
-matrix_multiply
-burst store_and_load 1 19 19
-burst store_and_load 19 16 2
-matrix_multiply
-burst store_and_load 2 19 19
-burst store_and_load 19 7 12
-nop
-matrix_multiply
-burst store_and_load 18 19 19
-burst store_and_load 19 1 2
-matrix_multiply
-burst store_and_load 8 19 19
-burst store_and_load 19 6 10
-nop
-matrix_multiply
-burst store_and_load 4 19 19
-burst store_and_load 19 9 14
-matrix_multiply
-burst store_and_load 0 19 19
-burst store_and_load 19 14 18
-matrix_multiply
-burst store_and_load 17 19 19
-burst store_and_load 19 6 16
-nop
-matrix_multiply
-burst store_and_load 2 19 19
-burst store_and_load 19 7 11
-nop
-matrix_multiply
-burst store_and_load 17 19 19
-burst store_and_load 19 9 16
-nop
-matrix_multiply
-burst store_and_load 3 19 19
-burst store_and_load 19 4 8
-nop
-matrix_multiply
-burst store_and_load 17 19 19
-burst store_and_load 19 4 8
-nop
-matrix_multiply
-burst store_and_load 10 19 19
-burst store_and_load 19 6 8
-matrix_multiply
-burst store_and_load 1 19 19
-burst store_and_load 19 2 13
-matrix_multiply
-burst store_and_load 0 19 19
-burst store_and_load 19 10 4
-matrix_multiply
-burst store_and_load 5 19 19
-burst store_and_load 19 14 17
-matrix_multiply
-burst store_and_load 3 19 19
-burst store_and_load 19 2 4
-matrix_multiply
-burst store_and_load 18 19 19
-burst store_and_load 19 17 4
-matrix_multiply
-burst store_and_load 9 19 19
-burst store_and_load 19 11 1
-matrix_multiply
-burst store_and_load 7 19 19
-burst store_and_load 19 3 11
-matrix_multiply
-burst store_and_load 4 19 19
-burst store_and_load 19 7 5
-matrix_multiply
-burst store_and_load 13 19 19
-burst store_and_load 19 0 5
-matrix_multiply
-burst store_and_load 13 19 19
-burst store_and_load 19 7 8
-nop
-matrix_multiply
-burst store_and_load 12 19 19
-burst store_and_load 19 1 15
-nop
-matrix_multiply
-burst store_and_load 11 19 19
-burst store_and_load 19 9 7
-nop
-matrix_multiply
-burst store_and_load 12 19 19
-burst store_and_load 19 10 8
-matrix_multiply
-burst store_and_load 11 19 19
-burst store_and_load 19 16 12
-matrix_multiply
-burst store_and_load 0 19 19
-burst store_and_load 19 3 8
-nop
-matrix_multiply
-burst store_and_load 1 19 19
-burst store_and_load 19 3 13
-matrix_multiply
-burst store_and_load 13 19 19
-burst store_and_load 19 16 3
-matrix_multiply
-burst store_and_load 8 19 19
+burst store_and_load 5 3 4
diff --git a/convert_data_from_data_in_plain_text_to_data_in.py b/convert_data_from_data_in_plain_text_to_data_in.py
index f85e3e4..4d2cf50 100644
--- a/convert_data_from_data_in_plain_text_to_data_in.py
+++ b/convert_data_from_data_in_plain_text_to_data_in.py
@@ -7,7 +7,7 @@ def convert_to_uppercase_hex_lines(input_file, output_file):
                 for val in numbers:
                     try:
                         num = int(val)
-                        hex_str = f"{num & 0xFFFF:04X}"
+                        hex_str = f"{num & 0xFFFFFFFF:08X}"
                         
                         f_out.write(hex_str + "\n")
                     except ValueError:
diff --git a/convert_data_from_data_out_to_data_out_plain_text.py b/convert_data_from_data_out_to_data_out_plain_text.py
index cfb11bb..7cdb0f6 100644
--- a/convert_data_from_data_out_to_data_out_plain_text.py
+++ b/convert_data_from_data_out_to_data_out_plain_text.py
@@ -9,8 +9,8 @@ def process_hex_data(input_file, output_file):
             for h in parts:
                 try:
                     val = int(h, 16)
-                    if val > (2**15 - 1):
-                        val -= (2**16)
+                    if val > (2**31 - 1):
+                        val -= (2**32)
                     decimals.append(val)
                     
                 except ValueError:
diff --git a/data_in_plain_text.txt b/data_in_plain_text.txt
index 581198e..466c2ba 100644
--- a/data_in_plain_text.txt
+++ b/data_in_plain_text.txt
@@ -1,23 +1,5 @@
-10 -7 -10 -2 -3 -3 -6 -7 7
--8 8 3 -9 -10 -8 -4 -3 6
-9 -10 7 -4 10 7 3 -3 4
-8 -2 -10 -5 3 0 -2 -6 -4
-0 -7 -8 2 -7 1 1 9 -2
--9 4 7 -7 2 -8 7 -1 10
-9 1 8 -4 -8 -9 -3 -1 -8
--3 -7 2 -2 4 10 1 -5 1
-1 -4 -2 10 -8 9 10 -5 7
--3 -5 4 2 -2 10 7 -3 0
--9 -3 -9 0 2 -2 -8 -4 8
-0 -4 10 5 2 10 4 -6 -2
--6 -3 7 7 -2 8 3 8 2
-1 -3 -6 6 5 -8 -9 -7 -6
-10 -5 3 9 -8 2 2 9 4
-6 -2 7 -10 -7 7 -2 10 0
--7 -1 3 -5 4 -10 -2 6 -5
-6 -7 10 -1 10 6 9 -4 -6
-1 -5 7 6 -10 9 0 5 -10
--7 1 -1 -3 -9 -3 8 -8 -8
+-64 -64 -64 -64 -64 -64 -64 -64 -64
+63 63 63 63 63 63 63 63 63
 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0
 0 0 0 0 0 0 0 0 0
diff --git a/src/tensor_core_controller.sv b/src/tensor_core_controller.sv
index 43acfd3..42e87ce 100644
--- a/src/tensor_core_controller.sv
+++ b/src/tensor_core_controller.sv
@@ -1,12 +1,6 @@
-`define NOP_OPCODE 2'b00
-`define TENSOR_CORE_OPCODE 2'b01
-`define BURST_OPCODE 2'b10
-
-`define BURST_STORE_SELECT 2'b00
-`define BURST_LOAD_SELECT 2'b01
-`define BURST_STORE_AND_LOAD_SELECT 2'b10
-`define BURST_MATRIX1_LOAD_SELECT 2'b11
-
+`define NOP_OPCODE 3'b000
+`define TENSOR_CORE_OPCODE 3'b001
+`define BURST_OPCODE 3'b010
 
 `define BUS_WIDTH 7
 
@@ -38,11 +32,10 @@ module tensor_core_controller (
     logic [3:0] burst_current_index; // stores the current index that the burst opcode is looking at either for storing or loading
     logic [`BUS_WIDTH:0] burst_load_negative_storage [2];
 
-    wire [1:0] burst_store_load_select;
-
     wire signed [`BUS_WIDTH:0] burst_current_dual_store_data [2];
     wire signed [`BUS_WIDTH:0] burst_current_quad_load_data [4];
     
+    
     wire [2:0] burst_quad_load_address;
     wire [`BUS_WIDTH:0] burst_quad_load_data [4];
 
@@ -52,13 +45,8 @@ module tensor_core_controller (
 
 
     wire [1:0] opcode;
-    wire [1:0] generic_opselect;
-    wire [2:0] operate_opselect;
 
-    assign opcode = current_instruction[1:0];
-    assign generic_opselect = current_instruction[3:2];
-    assign operate_opselect = current_instruction[4:2];
-    assign burst_store_load_select = current_instruction[3:2];
+    assign opcode = current_instruction[2:0];
 
     assign burst_current_quad_load_data[0] = burst_load_negative_storage[0];
     assign burst_current_quad_load_data[1] = burst_load_negative_storage[1];
@@ -82,7 +70,7 @@ module tensor_core_controller (
             is_burst_load_active <= 0;
         end
 
-        else if (opcode == `BURST_OPCODE && burst_store_load_select == `BURST_STORE_AND_LOAD_SELECT && (burst_current_index == 9 || burst_current_index == 8)) begin
+        else if (opcode == `BURST_OPCODE && (burst_current_index == 9 || burst_current_index == 8)) begin
             burst_current_index <= 0;
             is_burst_load_active <= 1;
             is_burst_store_active <= 1;
@@ -130,7 +118,7 @@ module tensor_core_controller (
         .clock_in(clock_in),
         .reset_in(reset_in),
 
-        .should_start_tensor_core(opcode == `TENSOR_CORE_OPCODE && (!is_burst_load_active || burst_current_index == 4)),
+        .should_start_tensor_core(opcode == `TENSOR_CORE_OPCODE && (!is_burst_load_active || burst_current_index == 8)),
 
         .tensor_core_input1(tensor_core_input1), .tensor_core_input2(tensor_core_input2),
         .tensor_core_output(tensor_core_output)
diff --git a/src/tensor_core_memory_controller.sv b/src/tensor_core_memory_controller.sv
index 30a0d5c..084b072 100644
--- a/src/tensor_core_memory_controller.sv
+++ b/src/tensor_core_memory_controller.sv
@@ -1,12 +1,7 @@
-`define NOP_OPCODE 2'b00
-`define TENSOR_CORE_OPERATE_OPCODE 2'b01
-`define BURST_OPCODE 2'b10
-`define RESET_OPCODE 2'b11
-
-`define BURST_STORE_SELECT 2'b00
-`define BURST_LOAD_SELECT 2'b01
-`define BURST_STORE_AND_LOAD_SELECT 2'b10
-
+`define NOP_OPCODE 3'b000
+`define TENSOR_CORE_OPERATE_OPCODE 3'b001
+`define BURST_OPCODE 3'b010
+`define RESET_OPCODE 3'b011
 
 `define BUS_WIDTH 7
 
@@ -22,38 +17,36 @@ module tensor_core_memory_controller(
 );
 
     logic [63:0] machine_code [0:20000];
-    logic [15:0] data [0:20000];
+    logic [31:0] data [0:20000];
     logic [15:0] current_machine_code_instruction_index_positive_edge;
     logic [15:0] current_machine_code_instruction_index_negative_edge;
     wire [15:0] current_machine_code_instruction_index;
 
-    logic scale_factor;
-
-
-    assign clock_out = clock_in;
 
     logic power_on_reset_signal = 1;
     logic positive_edge_reset_called;
 
     assign reset_out = (current_opcode == `RESET_OPCODE) || reset_in || power_on_reset_signal;
+    assign clock_out = clock_in;
+
 
 
 
     logic [63:0] raw_current_instruction;
-    wire [1:0] current_opcode;
-    wire [1:0] burst_store_load_select;
+    wire [2:0] current_opcode;
 
     assign raw_current_instruction = machine_code[current_machine_code_instruction_index];
-    assign current_opcode = raw_current_instruction[1:0];
-    assign burst_store_load_select = raw_current_instruction[3:2];
+    assign current_opcode = raw_current_instruction[2:0];
 
 
 
     wire is_burst_load_active;
     wire is_burst_store_active;
     
-    assign is_burst_load_active = ((burst_store_load_select == `BURST_STORE_AND_LOAD_SELECT || burst_store_load_select == `BURST_LOAD_SELECT) && current_opcode == `BURST_OPCODE && raw_current_instruction[15] == 1'b1);
-    assign is_burst_store_active = ((burst_store_load_select == `BURST_STORE_AND_LOAD_SELECT || burst_store_load_select == `BURST_STORE_SELECT) && current_opcode == `BURST_OPCODE && raw_current_instruction[14] == 1'b1);
+    wire [15:0] debug1;
+    assign debug1 = raw_current_instruction[15:0];
+    assign is_burst_load_active = (current_opcode == `BURST_OPCODE && raw_current_instruction[15] == 1'b1);
+    assign is_burst_store_active = (current_opcode == `BURST_OPCODE && raw_current_instruction[14] == 1'b1);
 
 
 
@@ -95,22 +88,18 @@ module tensor_core_memory_controller(
 
         // Process the current instruction word if it is some kind of load or reset instructions
         if (reset_out) begin
-            scale_factor = 0;
             current_tensor_core_instruction = 0;
         end
         else if (is_burst_load_active) begin
-            scale_factor = raw_current_instruction[11:4];
             current_tensor_core_instruction = {data[data_load_address1][`BUS_WIDTH:0], data[data_load_address2][`BUS_WIDTH:0]};
             
         end
 
         else if (is_burst_store_active) begin
-            scale_factor = raw_current_instruction[11:4];
             current_tensor_core_instruction = 0;
         end
 
         else begin
-            scale_factor = 0;
             current_tensor_core_instruction = raw_current_instruction[15:0];
         end
 
@@ -121,13 +110,17 @@ module tensor_core_memory_controller(
     // handle writing data that is being stored into main memory from the tensor core
     always_ff @(posedge clock_in) begin
         if (data_store_enable) begin
-            data[data_store_address][7:0] <= (data_store_data << scale_factor);
+            data[data_store_address][7:0] <= data_store_data;
         end
     end
 
     always_ff @(negedge clock_in) begin
         if (data_store_enable) begin
-            data[data_store_address][15:8] <= (data_store_data << scale_factor);
+            data[data_store_address][15:8] <= data_store_data;
+
+            if (data_store_data[7] == 1) begin
+                data[data_store_address][31:16] <= 16'hFFFF;
+            end
         end
     end
 
